### DISCUSSION TITLE
I'd like to retrieve the JSON for a Data List property. Could anyone walk me through it?

### DISCUSSION THREAD

**Discussion by markadrake** [leekelleher/umbraco-contentment#425] [Question] [answered]
I know the content type,
and I know the property alias, 
can I retrieve the JSON values for a specific Data List property?

(Screenshot below should help. I'm having difficulty phrasing my question.)

You might ask why I need them,
and it's because I need to use them in other locations, like an Umbraco Form, 
or list them somewhere on the front end,
use them outside of the main Umbraco Backoffice.

- Is there a way to query this through C#?
- Is there a way to query this through JS/API?

![image](https://github.com/user-attachments/assets/dd8c6e90-3c36-4837-a34e-3f7565dc8611)

**Comment by markadrake** [comment]
I discovered the `GetDataSourceItemsByDataTypeKey` endpoint. I think this is exactly what I need, but for some reason I'm only getting a 417 error back from the request. I've included all of the cookies with my `fetch` request.

> /umbraco/backoffice/Contentment/DataListApi/GetDataSourceItemsByDataTypeKey?key=f8351fcb-cace-431d-85d8-9b2f4c11d1ae

**(Previously...)**

With enough data, I may be able to use the `GetPreview` endpoint to retrieve these values.

> /umbraco/backoffice/Contentment/DataListApi/GetPreview


![image](https://github.com/user-attachments/assets/c5c97dcd-5ee9-479a-98ee-352ef74fce92)

![image](https://github.com/user-attachments/assets/9f0f2e6f-2a18-4ad0-bc98-12c4f1c272c7)

**Reply by leekelleher** [reply, marked as answer]
Hi @markadrake, the `DataListApiController` is restricted for backoffice auth/usage, so wouldn't be able to use on the frontend website.

The [guts of the `GetDataSourceItemsByDataTypeKey()` endpoint](https://github.com/leekelleher/umbraco-contentment/blob/5.0.2/src/Umbraco.Community.Contentment/DataEditors/DataList/DataListApiController.cs#L66-L76) could be used in your own controller/utility/helper method. You'd need to know the data-type ID/Guid.

**Reply by markadrake** [reply]
Thank you @leekelleher 

I just tried copying it to my controller, and it worked great!

I noticed `DataListDataEditor.DataEditorAlias` is internal (or maybe private). It's not preventing me from anything, but I feel more secure referencing strings from a model than what I did below!

Thanks for all the work you do and contribute to the Umbraco community!

```
    [HttpGet("possibleValues")]
    [ResponseCache(Location = ResponseCacheLocation.Any, Duration = 3600)]
    [ProducesResponseType(typeof(object), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(object), StatusCodes.Status404NotFound)]
    public object PossibleValues(Guid? key)
    {
        if (key.HasValue != true) 
            return NotFound();
        
        if (_dataTypeService.GetDataType(key.Value) is not { } dataType) 
            return NotFound();

        if(dataType?.EditorAlias.InvariantEquals("Umbraco.Community.Contentment.DataList") != true)
            return NotFound();

        if(dataType.Configuration is not Dictionary<string, object> config)
            return NotFound();

        if(_propertyEditors.TryGet(dataType.EditorAlias, out var propertyEditor) != true) 
            return NotFound();
        
        var configurationEditor = propertyEditor.GetConfigurationEditor();
        var valueEditorConfig = configurationEditor.ToValueEditor(config);

        return Ok(valueEditorConfig?["items"]);

    }
```

**Reply by markadrake** [reply]
Please ignore my overly verbose code. :)

**Reply by leekelleher** [reply]
It's undocumented, but there are a bunch of constants available for public use. They're all under a static `ContentmentConstants`, for simplicity (otherwise the constants would be scattered all over the place).

For the Data List alias can be accessed using: [`ContentmentConstants.PropertyEditors.Aliases.DataList`](https://github.com/leekelleher/umbraco-contentment/blob/5.0.2/src/Umbraco.Community.Contentment/ContentmentConstants.cs#L22).
